pkg/cmd/cli/cmd/set/triggers.go

```diff
@@ -0,0 +1,740 @@
+package set
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"reflect"
+	"strings"
+	"text/tabwriter"
+
+	"github.com/golang/glog"
+	"github.com/spf13/cobra"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/meta"
+	kcmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/kubectl/resource"
+	"k8s.io/kubernetes/pkg/runtime"
+
+	buildapi "github.com/openshift/origin/pkg/build/api"
+	buildutil "github.com/openshift/origin/pkg/build/util"
+	cmdutil "github.com/openshift/origin/pkg/cmd/util"
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+	deployapi "github.com/openshift/origin/pkg/deploy/api"
+	"github.com/openshift/origin/pkg/generate/app"
+	imageapi "github.com/openshift/origin/pkg/image/api"
+	"k8s.io/kubernetes/pkg/util/sets"
+)
+
+const (
+	triggersLong = `
+Set or remove triggers for build configs and deployment configs
+
+All build configs and deployment configs may have a set of triggers that result in a new deployment
+or build being created. This command enables you to alter those triggers - making them automatic or
+manual, adding new entries, or changing existing entries.
+
+Deployments support triggering off of image changes and on config changes. Config changes are any
+alterations to the pod template, while image changes will result in the container image value being
+updated whenever an image stream tag is updated.
+
+Build configs support triggering off of image changes, config changes, and webhooks (both GitHub-specific
+and generic). The config change trigger for a build config will only trigger the first build.`
+
+	triggersExample = `  # Print the triggers on the registry
+  $ %[1]s triggers dc/registry
+
+  # Set all triggers to manual
+  $ %[1]s triggers dc/registry --manual
+
+  # Enable all automatic triggers
+  $ %[1]s triggers dc/registry --auto
+
+  # Reset the GitHub webhook on a build to a new, generated secret
+  $ %[1]s triggers bc/webapp --from-github=
+  $ %[1]s triggers bc/webapp --from-webhook=
```

Comment:
Maybe --from-webhook should just be a boolean, and take no flags, and
generate a secret always.

On Wed, Feb 24, 2016 at 2:42 PM, Fabiano Franz <notifications@github.com>
wrote:

> In pkg/cmd/cli/cmd/set/triggers.go
> <https://github.com/openshift/origin/pull/7423#discussion_r53994859>:
>
> > +
> > +Build configs support triggering off of image changes, config changes, and webhooks (both GitHub-specific
> > +and generic). The config change trigger for a build config will only trigger the first build.`
> > +
> > +	triggersExample = `  # Print the triggers on the registry
> > +  $ %[1]s triggers dc/registry
> > +
> > +  # Set all triggers to manual
> > +  $ %[1]s triggers dc/registry --manual
> > +
> > +  # Enable all automatic triggers
> > +  $ %[1]s triggers dc/registry --auto
> > +
> > +  # Reset the GitHub webhook on a build to a new, generated secret
> > +  $ %[1]s triggers bc/webapp --from-github=
> > +  $ %[1]s triggers bc/webapp --from-webhook=
>
> What about this for the sake of usability (while making sure --from-thing=
> still works)
>
> --generate-github
> --generate-webhook
>
> --reset-github
> --reset-webhook
>
> I know, more flags, but setting something to empty string in order to
> generate a new secret looks like a strange pattern.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7423/files#r53994859>.
>